reports/models.py

```diff
@@ -0,0 +1,459 @@
+import logging
+from datetime import datetime
+from typing import List, Optional, Union
+
+import occurrences.models as occurrences_models
+import reports.services as reports_services
+from django.contrib.postgres.fields import ArrayField
+from django.db import models
+from django.utils.translation import ugettext_lazy as _
+from reports.exceptions import EnrolmentReportCouldNotHydrateLinkedEventsData
+from reports.utils import haversine
+
+from common.models import TimestampedModel
+
+logger = logging.getLogger(__name__)
+
+
+class EnrolmentQuerySet(models.QuerySet):
+    def unsynced(
+        self, sync_from: datetime = None
+    ) -> Union[models.QuerySet, List[occurrences_models.Enrolment]]:
+        """Get a list of enrolments which are updated after latest report updates."""
+        sync_from = sync_from or UnsyncedQuerySet.latest_sync(self)
+        if sync_from:
+            return occurrences_models.Enrolment.objects.filter(
+                updated_at__gte=sync_from
+            )
+        return occurrences_models.Enrolment.objects.all()
+
+    def missing(
+        self, sync_from: datetime = None
+    ) -> Union[models.QuerySet, List[occurrences_models.Enrolment]]:
+        """Get a list of enrolments which are missing from EnrolmentReport db-table."""
+        enrolments = EnrolmentQuerySet.unsynced(self, sync_from=sync_from)
+        reports_enrolment_ids = self.filter(
+            _enrolment_id__in=[e.id for e in enrolments]
+        ).values_list("_enrolment_id", flat=True)
+        return enrolments.exclude(id__in=reports_enrolment_ids)
+
+
+class UnsyncedQuerySet(models.QuerySet):
+    def latest_sync(self) -> Optional[datetime]:
+        """Returns latest updated_at, which represents the date of the last sync made"""
+        return self.aggregate(models.Max("updated_at"))["updated_at__max"]
+
+    def unsynced(self):
+        """
+        Query unsynced enrolment report instances:
+        Enrolment report is not in sync when it's related enrolment instance
+        has an updated_at -field value greater than
+        report instances updated_at -field value.
+        """
+        # Even if the enrolment instance has updated,
+        # we do not want to update it's related enrolment report instance,
+        # until the status of enrolment has changed.
+        enrolment_ids_with_statuses = list(
+            {"_enrolment_id": id, "enrolment_status": status}
+            for id, status in EnrolmentQuerySet.unsynced(self).values_list(
```

Comment:
If `EnrolmentQuerySet` is turned into functions this could be:

```suggestion
            for id, status in get_unsynched_enrollments().values_list(
```